temporian/implementation/numpy_cc/operators/custom_heap.h

```diff
@@ -0,0 +1,76 @@
+#include <iostream>
+#include <list>
+template <typename T>
+class CustomHeap {
+ private:
+  std::function<bool(T, T)> compare;
+  std::list<T> heap;
+  std::unordered_map<T, typename std::list<T>::iterator> val_to_node;
+
+ public:
+  CustomHeap(std::function<bool(T, T)> compare) : compare(compare) {}
+
+  void push(T value) {
+    heap.push_back(value);
+    auto it = std::prev(heap.end());
+    val_to_node[value] = it;
+    // TODO: better sorting?
+    while (it != heap.begin()) {
+      auto parent = std::prev(it);
+      if (!compare(*parent, *it)) {
+        break;
+      }
+      // TODO: check that this swap is doing what I want
+      std::swap(*parent, *it);
+      val_to_node[*it] = it;
+      val_to_node[*parent] = parent;
+      it = parent;
+    }
+  }
+
+  std::optional<T> pop() {
+    if (heap.size() == 0) {
+      return {};
+    } else {
+      auto value = heap.back();
+      heap.pop_back();
+      auto it = val_to_node.find(value);
```

Comment:
What if there are multiple time the same value?

Reply:
That would be a problem I forgot to add a comment addressing that. It's not a problem in this case because I'm storing the indices as the heap "values" so they are never going to repeat